demo/src/components/Cursors.tsx

```diff
@@ -1,127 +1,91 @@
-import { useContext, useEffect, useReducer } from 'react';
-import assign from 'lodash.assign';
+import { useContext, useEffect, useState } from 'react';
+import type { CursorUpdate as _CursorUpdate } from '@ably/spaces';
+
 import cn from 'classnames';
-import find from 'lodash.find';
-import omit from 'lodash.omit';
 import { CursorSvg, SpacesContext } from '.';
 import { useMembers, CURSOR_ENTER, CURSOR_LEAVE, CURSOR_MOVE } from '../hooks';
-import { type Member } from '../utils/types';
-
-type ActionType = 'move' | 'enter' | 'leave';
-
-interface Action {
-  type: ActionType;
-  data: {
-    connectionId: string;
-    members?: Member[];
-    position?: {
-      x: number;
-      y: number;
-    };
-  };
-}
 
-interface State {
-  [connectionId: string]: Member & Action['data'];
-}
-
-const reducer = (state: State, action: Action): State => {
-  const { type, data } = action;
-  const { connectionId, members, position } = data;
-  switch (type) {
-    case CURSOR_ENTER:
-      return {
-        ...state,
-        [connectionId]: {
-          ...assign(find(members, { connectionId }), { connectionId, position }),
-        },
-      };
-    case CURSOR_LEAVE:
-      return {
-        ...omit(state, connectionId),
-      };
-    case CURSOR_MOVE:
-      return {
-        ...state,
-        [connectionId]: {
-          ...assign(find(members, { connectionId }), { connectionId, position }),
-        },
-      };
-    default:
-      throw new Error('Unknown dispatch type');
-  }
-};
+type state = typeof CURSOR_ENTER | typeof CURSOR_LEAVE | typeof CURSOR_MOVE;
+type CursorUpdate = Omit<_CursorUpdate, 'data'> & { data: { state: state } };
 
 export const Cursors = () => {
   const space = useContext(SpacesContext);
-  const { self, members } = useMembers();
-  const [activeCursors, dispatch] = useReducer(reducer, {});
+  const { self, others } = useMembers();
+  const [cursors, setCursors] = useState<{
+    [connectionId: string]: { position: CursorUpdate['position']; state: CursorUpdate['data']['state'] };
+  }>({});
 
   useEffect(() => {
-    if (!space || !members) return;
+    if (!space || !others) return;
 
     space.cursors.subscribe('update', (cursorUpdate) => {
-      const { connectionId } = cursorUpdate;
-      const member = find<Member>(members, { connectionId });
+      const { connectionId, position, data } = cursorUpdate as CursorUpdate;
```

Comment:
What's the purpose of this type assertion? I'm guessing that the idea is to narrow the type of `cursorUpdate.data`. But it means that we throw away all of the other type information related to `cursorUpdate` (i.e. all of the non-`data` properties of `_CursorUpdate`), which does not seem desirable.

Reply:
Sorry, I think I wasn't very clear. What I mean is that by introducing a type assertion, we're telling TypeScript to disregard the information that it knows about the type of the object that `space.cursors.subscribe` passes to its callback.

That is, we're telling it that it should forget about the properties that it knows to exist on `cursorUpdate`, and to instead trust that it will contain the properties described by the `CursorUpdate` type that we've defined above. For example, we could add a new property to `CursorUpdate` — one which doesn’t exist on the `_CursorUpdate` type that we've imported from the `@ably/spaces` — and it will not cause a compilation error:

```typescript
type CursorUpdate = Omit<_CursorUpdate, 'data'> & { foo: string } & { data: { state: state } };
```

Reply:
My suggestion would be that we only apply the type assertion to the `data` property; that is, get rid of the `CursorUpdate` type and do the following (there might be a neater way of doing it but you get the idea):

```typescript
const { connectionId, position } = cursorUpdate;
const data = cursorUpdate.data as { state: state };
```

Reply:
I agree that your suggestion is more explicit, but it still depends on `as`. As soon as we use it, we override the checker and the effect is the same `{ state: state }` could not exist at all. As such, I don't really see a point in the change. A proper improvement for me would be to not have to use `as` which I think can be achieved in the future with generics.

Reply:
No, it's not the same. In my example, the `as` is being used in a very targeted way, to only provide type information for the object returned by `cursorUpdate`’s `data` property, which is initially of type `unknown`. Whereas, in your code, the type assertion is providing type information for the entire `cursorUpdate` object.

Yes, you are correct that my type assertion could lead to the introduction of non-existent properties on `data`. But there's no way around that, given that `cursorUpdate.data` is of type `unknown`. The difference is that your code is essentially treating _the entirety of the `cursorUpdate` object_ as if it were of type `unknown`, when in fact we have plenty of strong type data about the properties that exist on that object. Does that make sense?

I also am not really sure how generics would help us here. I'm assuming that you're suggesting that we make Spaces’ `CursorData` type generic, parameterised on the type of the `data` object? That doesn't really help, though, since it doesn't get around the fact the user is going to, at some point, have to somehow tell the compiler "I know that you don't know what type is contained in this cursor data’s `data` property, but trust me, it's of type `Foo`".

Reply:
> No, it's not the same. In my example, the as is being used in a very targeted way

Not disagreeing here, just saying we are still overriding the compiler, that's all. The goal is not to use `as` and I'm convinced that's possible. Additionally, the way this is written now, I can reuse the type, here, your way is mostly convenient because we do it once. In other examples we've written, you'd need to write `as` in every single subscriber. 

I am still basing my type on the type that comes from the Spaces - what would be the point of exporting a `CursorUpdate` here if we can't trust it?

> That doesn't really help, though, since it doesn't get around the fact the user is going to, at some point, have to somehow tell the compiler

Let's chat about this separately :) For the purpose of this PR I will leave it as it is because I'm not convinced your suggestion is impactful enough (and would diverge from a pattern we already have in many other examples), but I definitely want something better for GA.

Reply:
Whilst I personally believe that there is a meaningful difference between the two, I'm happy to go with your judgement on it. I'll resolve this conversation.